### PR TITLE
FEATURE:  F1-277 design and data transform for forecast, execution

### PR DESCRIPTION
**Pull request by hackerstanislav**
JIRA: F1-277

---

> [!IMPORTANT]
> Please, **don't forget to run `rush change`** for the commits that introduce **new features** 🙏

---

Refer to [documentation](https://github.com/gooddata/gooddata-ui-sdk/blob/master/dev_docs/continuous_integration.md) to see how to run checks and tests in the pull request. This is the list of the most used commands:

```
extended test - backstop
```

```
extended test - tiger-cypress - integrated
extended test - tiger-cypress - isolated
extended test - tiger-cypress - record
```

### CODE REVIEW

**Comment by hackerstanislav** [comment]
extended test - backstop

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/18871/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/20594/ : SUCCESS in 19m 53s

**Comment by gdgate** [comment]
Build in pipeline check-extended aborted.

**Comment by hackerstanislav** [comment]
extended test - backstop

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/18873/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/20595/ : SUCCESS in 10m 11s

File: libs/sdk-backend-base/src/cachingBackend/index.ts [hunk 1]
```diff
@@ -224,6 +228,19 @@ class WithExecutionResultCaching extends DecoratedExecutionResult {
         return this.allData;
     };
 
+    public readForecastAll = (config: IForecastConfig): Promise<IForecastResult> => {
+        // TODO: enable forecasting caching size configuration
```
Comment (kandl) [inline libs/sdk-backend-base/src/cachingBackend/index.ts:232]: Should we create a ticket for this? Or handle it right now?

File: libs/sdk-backend-spi/src/workspace/execution/index.ts [hunk 2]
```diff
@@ -293,6 +322,18 @@ export interface IExecutionResult {
      */
     readWindow(offset: number[], size: number[]): Promise<IDataView>;
 
+    /**
+     * Reads forecast for the execution result.
+     * @beta
+     */
+    readForecastAll(config: IForecastConfig): Promise<IForecastResult>;
+
+    // TODO: add support for forecast paging
```
Comment (kandl) [inline libs/sdk-backend-spi/src/workspace/execution/index.ts:331]: Probably lets remove this, implement it, or create a ticket for it.

File: libs/sdk-backend-spi/src/workspace/execution/index.ts [hunk 3]
```diff
@@ -456,4 +509,31 @@ export interface IDataView {
      * Thus, two data views on the same result, with same offset and limit will have the same fingerprint.
      */
     fingerprint(): string;
+
+    /**
+     * Calculate forecast data
+     * @beta
+     */
+    forecast(): IForecastView;
```
Comment (kandl) [inline libs/sdk-backend-spi/src/workspace/execution/index.ts:517]: This is not true - it does not calculate anything (forecast config + result is provided by `withForecast` method). Also what about if the forecast for this DataView was not provided? Should it throw, or return undefined?

File: libs/sdk-ui-charts/api/sdk-ui-charts.api.md [hunk 4]
```diff
@@ -524,6 +529,14 @@ export interface IDonutChartBucketProps {
 export interface IDonutChartProps extends IBucketChartProps, IDonutChartBucketProps {
 }
 
+// @beta
+export interface IForecast {
+    confidence: "70" | "75" | "80" | "85" | "90" | "95" | number;
```
Comment (kandl) [inline libs/sdk-ui-charts/api/sdk-ui-charts.api.md:534]: Does it need to be string numbers? I would either type it as number, or union of allowed numbers.

**Comment by yenkins** [comment]
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/2638/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/18874/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/2149/ : SUCCESS in 24m 36s

File: libs/sdk-backend-base/src/cachingBackend/index.ts (hunk 1 above)
Comment (hackerstanislav) [reply libs/sdk-backend-base/src/cachingBackend/index.ts:232]: https://gooddata.atlassian.net/browse/F1-305

File: libs/sdk-backend-spi/src/workspace/execution/index.ts (hunk 2 above)
Comment (hackerstanislav) [reply libs/sdk-backend-spi/src/workspace/execution/index.ts:331]: I will remove for now, hope will not be neccesary for now

File: libs/sdk-ui-charts/api/sdk-ui-charts.api.md (hunk 4 above)
Comment (hackerstanislav) [reply libs/sdk-ui-charts/api/sdk-ui-charts.api.md:534]: This is special case to have better typecheck in config panel. From config panel we can use only these string values, in API of chart config we can use also numbers. The logic is try to parse nubmer from string, normalize it and convert to 0 - 1 finally. I can remove it, but it stop completion for config panels. We can keep there only number and string.

File: libs/sdk-ui-charts/api/sdk-ui-charts.api.md [hunk 5]
```diff
@@ -524,6 +529,14 @@ export interface IDonutChartBucketProps {
...
+    enabled: boolean;
+    period: number | string;
```
[the remark this answers is not in the document]
Comment (hackerstanislav) [reply libs/sdk-ui-charts/api/sdk-ui-charts.api.md:536]: This is special case to have better typecheck in config panel. From config panel we can use only these string values, in API of chart config we can use also numbers. The logic is try to parse nubmer from string, normalize it and convert to 0 - 1 finally. I can remove it, but it stop completion for config panels. We can keep there only number and string.

File: libs/sdk-ui-charts/src/forecast.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    enabled: boolean;
+    visible: boolean;
+} {
+    switch (type) {
+        case "line": {
+            const buckets = insightBuckets(insight);
+            const measures = bucketsFind(buckets, (b) => b.localIdentifier === BucketNames.MEASURES);
+            const trends = bucketsFind(buckets, (b) => b.localIdentifier === BucketNames.TREND);
+
+            //TODO: s.hacker check if the trend is date attribute somehow
```
[the remark this answers is not in the document]
Comment (hackerstanislav) [reply libs/sdk-ui-charts/src/forecast.ts:23]: https://gooddata.atlassian.net/browse/F1-306

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/18875/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/20596/ : FAILURE in 1m 17s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/18876/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/20597/ : SUCCESS in 10m 07s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/18886/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/20607/ : SUCCESS in 17m 54s

**Review by kandl — APPROVED** [review]
LGTM

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul/2138/ : SUCCESS in 1m 13s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/20608/ : SUCCESS in 18m 34s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-tiger-isolated-zuul-pipeline/1937/ : SUCCESS in 10m 52s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-bear-isolated-zuul-pipeline/1917/ : SUCCESS in 6m 48s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/5435/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-compatibility-pipeline/740/ : SUCCESS in 5m 45s

**Pull request merged**